runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -129,82 +123,109 @@ public long getTail() {
-     */
-    private void removeAddresses(List<Long> addresses) {
-        addresses.stream().forEach(bitmap::removeLong);
-
-        // Recover allocated but unused memory
-        bitmap.trim();
-        // TODO(Maithem) if addresses is empty, the trim mark will not be updated
-        trimMark = Collections.max(addresses);
-
-        log.trace("removeAddresses: new trim mark set to {}", trimMark);
-    }
-
     /**
      * Trim all addresses lower or equal to trimMark and set new trim mark.
      *
-     * @param trimMark upper limit of addresses to trim
+     * @param newTrimMark upper limit of addresses to trim
      */
-    public void trim(long trimMark) {
-        if (!Address.isAddress(trimMark)) {
+    public void trim(long newTrimMark) {
+        if (Address.nonAddress(newTrimMark)) {
             // If not valid address return and do not attempt to trim.
-            if (log.isTraceEnabled()) {
-                log.trace("trim: attempting to trim non-valid address {}", trimMark);
-            }
+            log.trace("trim: attempting to trim non-valid address {}", newTrimMark);
+            return;
+        }
+
+        // If the new trim mark is greater than the current trim mark it needs to be
+        // updated regardless whether it will result in addresses being removed or not.
+        // For example, the bitmap can be empty, while the trim mark keeps increasing
+        if (newTrimMark <= this.trimMark) {
             return;
         }
 
-        // Note: if a negative value is passed to this API the cardinality
-        // of the bitmap is returned, which would be incorrect as we would
-        // be removing all addresses upon an invalid trim mark.
-        long numAddressesToTrim = bitmap.rankLong(trimMark);
+        // At this point point the trim mark is positive and we need to compute
+        // the rank. The rank will indicate how many addresses need to be trimmed
+        long rank = bitmap.rankLong(newTrimMark);
 
-        if (numAddressesToTrim <= 0) {
+        if (rank <= 0) {
             return;
         }
 
-        List<Long> addressesToTrim = new ArrayList<>();
-        LongIterator it = bitmap.getLongIterator();
-        for (int i = 0; i < numAddressesToTrim; i++) {
-            addressesToTrim.add(it.next());
+        Roaring64NavigableMap trimmedBitmap = new Roaring64NavigableMap();
+        LongIterator iterator = this.bitmap.getReverseLongIterator();
+        while (iterator.hasNext()) {
+            long current = iterator.next();
+            if (current <= newTrimMark) {
+                log.info("trim: stream trim mark moved from {} to {}", this.trimMark, current);
+                this.trimMark = current;
+                break;
+            }
+
+            trimmedBitmap.addLong(current);
         }
 
-        log.trace("trim: Remove {} addresses for trim mark {}", addressesToTrim.size(), trimMark);
+        this.bitmap = trimmedBitmap;
```

Comment:
Let's say before the trim, address is [trimMark + 1, max]. Do we assume that [trimMark + 1, new trimMark] always has more addresses than [new trimMark + 1, max]?

If not, is it possible to calculate which part has less addresses? If [trimMark + 1, new trimMark] is smaller, is removing address better in this case?

Reply:
I thought about this when I first implemented it and I couldn't find a good general solution. The problem is to find that point where it's cheaper to removes VS when to copy is an expensive series of operations. For example, you need to do a couple of rank operations and then do the copy/remove, when I tested it a copy of the suffix turned out to be faster. Trim is called infrequently too, I think just paying the cost of the copy is reasonable and doesn't seem to require optimizing. 

If you have any suggestions, I can try them too.